README.md
# WeNet - Profile Diversity Manager


## Introduction

This project is an extension of the profile manager to allow it to calculate
the similarity between strings.

The component to manage the diversity of the user profiles.

## Setup and configuration

First of all, you must install the next software.

 - [docker](https://docs.docker.com/install/)
 - [docker compose](https://docs.docker.com/compose/install/)

### Development

First of all you need to installl the [docker](https://docker.io),
and after that you can run the script **startDevelopmentEnvironment.sh**
to start the development environment.
On this environment you have access to the commands:

 - **start-reload.sh** start a development server that detect the changes
  on the files and start again.
 - **start.sh** start a server similar to the production one. This server
  Not detect the changes, so to be effective you must stop and start again.
 
In both casses you can interact with the stertaed server
at **http://localhost/doc** or at **http://localhost/redoc**.
Also you can stop the server with keys combinations **Control-C**.

If you go to the **tests** directory and execute the command `pytest --cov=../app`
to run all test and obtain a coverage measure of them.

If you not **exit** from the development environment, may be the next time
you try to start it will fail. In this case before to start again try
the next commands:

```
docker stop profile_diversity_manager_dev
docker rm profile_diversity_manager_dev
```

### Create docker image

You can create the docker container with the script **buildDockerImage.sh**.
The created container has the next parameters:

 - **HOST** it is the host to bind the server. By default is set to **0.0.0.0**.
 - **PORT** it is set port to bind teh server. By default is set to **80**.
 - **MAX_EXECUTOR_WORKERS** it is set to the number of backgound threads that
   the server will use. By default is set to **4**.
 - **LOG_LEVEL** the logging level (debug,info,warning,error or critical).
   By default is set to **info**.
 
   
Also you can use the environment veriables of [fatsapi base docker image](https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker).

After that you can run the service with the command: 

```
docker run -t internetofus/profile-diversity-manager:VERSION
```

Where **VERSION** is the version of the **profile-diversity-manager** you want to run.


## License

This software is under the [Apache V2 license](LICENSE)


## Interaction with other WeNet components

## Contact

### Researcher

 - [Nardine Osman](http://www.iiia.csic.es/~nardine/) ( [IIIA-CSIC](https://www.iiia.csic.es/~nardine/) ) nardine (at) iiia.csic.es
 - [Carles Sierra](http://www.iiia.csic.es/~sierra/) ( [IIIA-CSIC](https://www.iiia.csic.es/~sierra/) ) sierra (at) iiia.csic.es

### Developers

 - Joan Jené ( [UDT-IA, IIIA-CSIC](https://www.iiia.csic.es/people/person/?person_id=19) ) jjene (at) iiia.csic.es
 - Bruno Rosell i Gui ( [UDT-IA, IIIA-CSIC](https://www.iiia.csic.es/people/person/?person_id=27) ) rosell (at) iiia.csic.es


CHANGELOG
# Changelog

All notable changes to `wenet-profile-diversity-manager` will be documented in this file

## Version 0.2.0 (2021-11-9)

 * Add normalization for the profile attributes.
 

## Version 0.1.0 (2021-10-21)

 * Provide diversity calculus web service.
 * Provide similarity calculus web service.

tests/test_wns.py
#
# Copyright 2019 - 2022 UDT-IA, IIIA-CSIC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import pytest
from WNS import sim_str_str, sim_str_str_multiling, sim_str_attrlst, sim_str_attrlst_multiling


@pytest.mark.timeout(30)
def test_sim_str_str_es_es():
    """Test the similarity between two strings"""
    sim = sim_str_str("Odio los aviones", "No me gustan los helicópteros", lang1="es", lang2="es", stat="max")
    assert 0.0 <= sim <= 1.0


@pytest.mark.timeout(30)
def test_sim_str_str_zh_zh():
    """Test the similarity between two strings"""
    sim = sim_str_str("支持四种分词模式", "支持四种分词模式", lang1="zh", lang2="zh", stat="max")
    assert 0.0 <= sim <= 1.0


@pytest.mark.timeout(30)
def test_sim_str_str_multiling():
    """Test the similarity between two strings detecting language before"""
    sim = sim_str_str_multiling("I hate planes", "No me gustan los helicópteros", stat="max")
    assert 0.0 <= sim <= 1.0


@pytest.mark.timeout(30)
def test_sim_str_attrlst_es_es():
    """Test the similarity between a string an a set of attributes"""
    pairs = [("profesión", "programador aplicaciones"), ("afición", "escalada")]
    sims = sim_str_attrlst("Estoy probando esta nueva aplicación en mi portátil conectado a internet.", pairs, lang1="es", lang2="es", stat="max")
    assert len(pairs) == len(sims)
    for pair in pairs:
        for sim in sims:
            if pair[0] == sim[0]:
                assert pair[1] == sim[1]
                assert 0.0 <= sim[2] <= 1.0
                break;


@pytest.mark.timeout(30)
def test_sim_str_attrlst_multiling():
    """Test the similarity between a string an a set of attributes detecting language"""
    pairs = [("occupation", "app developer"), ("hobby", "climbing")]
    sims = sim_str_attrlst_multiling("Estoy probando esta nueva aplicación en mi portátil conectado a internet.", pairs, stat="max")
    assert len(pairs) == len(sims)
    for pair in pairs:
        for sim in sims:
            if pair[0] == sim[0]:
                assert pair[1] == sim[1]
                assert 0.0 <= sim[2] <= 1.0
                break;


tests/test_main.py
#
# Copyright 2019 - 2022 UDT-IA, IIIA-CSIC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import json
import pytest
import uuid

from fastapi.testclient import TestClient
from main import app
from pytest_httpserver import HTTPServer

client = TestClient(app)


@pytest.mark.timeout(30)
def test_get_help_info():
    """Test get help info of the API"""
    response = client.get("/help/info")
    assert response.status_code == 200
    body = response.json()
    assert 'name' in body
    assert 'apiVersion' in body
    assert 'softwareVersion' in body
    assert 'vendor' in body
    assert 'license' in body


@pytest.mark.timeout(30)
def test_post_calculate_diversity_of():
    """Test calculate diversity of two agents"""
    data = {
        "agents":[
                {
                    "id":"1",
                    "quantitativeAttributes":{
                            "introvert":1.0,
                            "extrovert":1.0,
                            "naturalist":1.0
                    },
                    "qualitativeAttributes":{
                            "gender":"M",
                            "civilStatus":"married"
                    }
                },
                {
                    "id":"2",
                    "quantitativeAttributes":{
                            "introvert":0.0,
                            "extrovert":0.0,
                            "naturalist":0.0
                    },
                    "qualitativeAttributes":{
                            "gender":"F",
                            "civilStatus":"single"
                    }
                }
            ],
        "qualitativeAttributes": {
                "gender":["M", "F", "O"],
                "civilStatus":["single", "married", "divorced", "widow"]
            },
        "quantitativeAttributes": ["introvert", "extrovert", "naturalist"]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.5261859507142914}


@pytest.mark.timeout(30)
def test_post_calculate_diversity_of_none_agents():
    """Test calculate diversity of none agents"""
    data = {
        "qualitativeAttributes": {
                "gender":["M", "F", "O"],
                "civilStatus":["single", "married", "divorced", "widow"]
            },
        "quantitativeAttributes": ["introvert", "extrovert", "naturalist"]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.0}


@pytest.mark.timeout(30)
def test_post_calculate_diversity_of_none_attribues():
    """Test calculate diversity of none attributes"""
    data = {
        "agents":[
                {
                    "id":"1",
                    "quantitativeAttributes":{
                            "introvert":1.0,
                            "extrovert":1.0,
                            "naturalist":1.0
                    },
                    "qualitativeAttributes":{
                            "gender":"M",
                            "civilStatus":"married"
                    }
                },
                {
                    "id":"2",
                    "quantitativeAttributes":{
                            "introvert":0.0,
                            "extrovert":0.0,
                            "naturalist":0.0
                    },
                    "qualitativeAttributes":{
                            "gender":"F",
                            "civilStatus":"single"
                    }
                }
            ]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.0}


@pytest.mark.timeout(30)
def test_post_calculate_similarity_of():
    """Test similarity some attributes"""
    data = {
        "source":"Do you have a bike?",
        "attributes":["car", "plane", "vehicle" ]
        }
    response = client.post("/calculateSimilarityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"similarities":[{"attribute":"car", "similarity":0.6239031848677311}, {"attribute":"plane", "similarity":0.0}, {"attribute":"vehicle", "similarity":0.6239031848677311}]}


@pytest.mark.timeout(30)
def test_post_calculate_similarity_of_with_non_normalized_attributes():
    """Test similarity some non normalized attributes"""
    data = {
        "source":"Do you have a bike?",
        "attributes":["materials.sportiveCar", "materials.bi-plane", "materials.undefined_vehicle" ]
        }
    response = client.post("/calculateSimilarityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"similarities":[{"attribute":"materials.sportiveCar", "similarity":0.6239031848677311}, {"attribute":"materials.bi-plane", "similarity":0.0}, {"attribute":"materials.undefined_vehicle", "similarity":0.6239031848677311}]}


@pytest.mark.timeout(30)
def test_post_calculate_similarity_of_none_attributes():
    """Test similarity without attributes"""
    data = {
        "source":"Do you have a bike?"
        }
    response = client.post("/calculateSimilarityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"similarities":[]}


@pytest.mark.timeout(30)
def test_post_calculate_similarity_of_empty_json_fail():
    """Test similarity with empty json fail"""
    data = {        }
    response = client.post("/calculateSimilarityOf", json=data)
    assert response.status_code == 422


@pytest.mark.timeout(30)
def test_post_calculate_diversity_for_domain_studying_career():
    """Test calculate diversity for domain studying career"""
    data = {
        "agents":[
                {
                    "id":"1",
                    "quantitativeAttributes":{
                        "competences.u_active":0.0,
                        "competences.u_read":0.0,
                        "competences.u_essay":0.0,
                        "competences.u_org":0.0,
                        "competences.u_balance":0.0,
                        "competences.u_assess":0.0,
                        "competences.u_theory":0.0,
                        "competences.u_pract":0.0                    
                    },
                    "qualitativeAttributes":{}
                },
                {
                    "id":"2",
                    "quantitativeAttributes":{
                        "competences.u_active":0.5,
                        "competences.u_read":0.5,
                        "competences.u_essay":0.5,
                        "competences.u_org":0.5,
                        "competences.u_balance":0.5,
                        "competences.u_assess":0.5,
                        "competences.u_theory":0.5,
                        "competences.u_pract":0.5                    
                    },
                    "qualitativeAttributes":{}
                }
            ]
        , "quantitativeAttributes": [
            "competences.u_active",
            "competences.u_read",
            "competences.u_essay",
            "competences.u_org",
            "competences.u_balance",
            "competences.u_assess",
            "competences.u_theory",
            "competences.u_pract"
            ]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.25}


@pytest.mark.timeout(30)
def test_post_calculate_diversity_for_domain_studying_career_2():
    """Test calculate diversity for domain studying career 2"""
    data = {
        "agents":[
                {
                    "id":"2",
                    "quantitativeAttributes":{
                        "competences.u_active":0.25,
                        "competences.u_read":0.25,
                        "competences.u_essay":0.25,
                        "competences.u_org":0.25,
                        "competences.u_balance":0.25,
                        "competences.u_assess":0.25,
                        "competences.u_theory":0.25,
                        "competences.u_pract":0.25                    
                    },
                    "qualitativeAttributes":{}
                },
                {
                    "id":"1",
                    "quantitativeAttributes":{
                        "competences.u_active":0.0,
                        "competences.u_read":0.0,
                        "competences.u_essay":0.0,
                        "competences.u_org":0.0,
                        "competences.u_balance":0.0,
                        "competences.u_assess":0.0,
                        "competences.u_theory":0.0,
                        "competences.u_pract":0.0                    
                    },
                    "qualitativeAttributes":{}
                }
            ]
        , "quantitativeAttributes": [
            "competences.u_active",
            "competences.u_read",
            "competences.u_essay",
            "competences.u_org",
            "competences.u_balance",
            "competences.u_assess",
            "competences.u_theory",
            "competences.u_pract"
            ]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.125}


@pytest.mark.timeout(30)
def test_post_calculate_diversity_for_domain_studying_career_3():
    """Test calculate diversity for domain studying career 3"""
    data = {
        "agents":[
                {
                    "id":"2",
                    "quantitativeAttributes":{
                        "competences.u_active":1,
                        "competences.u_read":1,
                        "competences.u_essay":1,
                        "competences.u_org":1,
                        "competences.u_balance":1,
                        "competences.u_assess":1,
                        "competences.u_theory":1,
                        "competences.u_pract":1                    
                    },
                    "qualitativeAttributes":{}
                },
                {
                    "id":"1",
                    "quantitativeAttributes":{
                        "competences.u_active":0.0,
                        "competences.u_read":0.0,
                        "competences.u_essay":0.0,
                        "competences.u_org":0.0,
                        "competences.u_balance":0.0,
                        "competences.u_assess":0.0,
                        "competences.u_theory":0.0,
                        "competences.u_pract":0.0                    
                    },
                    "qualitativeAttributes":{}
                }
            ]
        , "quantitativeAttributes": [
            "competences.u_active",
            "competences.u_read",
            "competences.u_essay",
            "competences.u_org",
            "competences.u_balance",
            "competences.u_assess",
            "competences.u_theory",
            "competences.u_pract"
            ]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.5}


@pytest.mark.timeout(30)
def test_post_calculate_diversity_for_domain_studying_career_4():
    """Test calculate diversity for domain studying career 4"""
    data = {
        "agents":[
                {
                    "id":"2",
                    "quantitativeAttributes":{
                        "competences.u_active":0,
                        "competences.u_read":0,
                        "competences.u_essay":0,
                        "competences.u_org":0,
                        "competences.u_balance":0,
                        "competences.u_assess":0,
                        "competences.u_theory":0,
                        "competences.u_pract":0                    
                    },
                    "qualitativeAttributes":{}
                },
                {
                    "id":"1",
                    "quantitativeAttributes":{
                        "competences.u_active":0.75,
                        "competences.u_read":0.75,
                        "competences.u_essay":0.75,
                        "competences.u_org":0.75,
                        "competences.u_balance":0.75,
                        "competences.u_assess":0.75,
                        "competences.u_theory":0.75,
                        "competences.u_pract":0.75                    
                    },
                    "qualitativeAttributes":{}
                }
            ]
        , "quantitativeAttributes": [
            "competences.u_active",
            "competences.u_read",
            "competences.u_essay",
            "competences.u_org",
            "competences.u_balance",
            "competences.u_assess",
            "competences.u_theory",
            "competences.u_pract"
            ]
        }
    response = client.post("/calculateDiversityOf", json=data)
    assert response.status_code == 200
    assert response.json() == {"value":0.375}


tests/test_diversity.py
#
# Copyright 2019 - 2022 UDT-IA, IIIA-CSIC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import pytest
from diversity import Agent, Attribute, QN, diversity


@pytest.mark.timeout(30)
def test_calculate_diversity_for_oposed_agents():
    """Test the diversity over two equals agents"""
    agents = [
        Agent("1234", [
            Attribute("property1", QN, [0, 1], 1),
            Attribute("property2", QN, [0, 1], 1),
            Attribute("property3", QN, [0, 1], 1)
            ]),
        Agent("abcd", [
            Attribute("property1", QN, [0, 1], 0),
            Attribute("property2", QN, [0, 1], 0),
            Attribute("property3", QN, [0, 1], 0)
            ])
    ]
    reqattr = [Attribute("property1", QN, [0, 1]), Attribute("property2", QN, [0, 1]), Attribute("property3", QN, [0, 1])]
    div = diversity(agents, reqattr)
    assert div == 1.5


@pytest.mark.timeout(30)
def test_calculate_diversity_for_equals_agents():
    """Test the diversity over two equals agents"""
    agents = [
        Agent("1234", [
            Attribute("property1", QN, [0, 1], 1),
            Attribute("property2", QN, [0, 1], 1),
            Attribute("property3", QN, [0, 1], 1)
            ]),
        Agent("abcd", [
            Attribute("property1", QN, [0, 1], 1),
            Attribute("property2", QN, [0, 1], 1),
            Attribute("property3", QN, [0, 1], 1)
            ])
    ]
    reqattr = [Attribute("property1", QN, [0, 1]), Attribute("property2", QN, [0, 1]), Attribute("property3", QN, [0, 1])]
    div = diversity(agents, reqattr)
    assert  div == 0.0


@pytest.mark.timeout(30)
def test_calculate_diversity_for_different_agents():
    """Test the diversity over two equals agents"""
    agents = [
        Agent("1234", [
            Attribute("property1", QN, [0, 1], 0.1),
            Attribute("property2", QN, [0, 1], 1),
            Attribute("property3", QN, [0, 1], 0.91)
            ]),
        Agent("abcd", [
            Attribute("property1", QN, [0, 1], 0.742),
            Attribute("property2", QN, [0, 1], 0.67),
            Attribute("property3", QN, [0, 1], 0.28)
            ])
    ]
    reqattr = [Attribute("property1", QN, [0, 1]), Attribute("property2", QN, [0, 1]), Attribute("property3", QN, [0, 1])]
    div = diversity(agents, reqattr)
    assert 0.0 <= div <= 3.0


@pytest.mark.timeout(30)
def test_calculate_diversity_for_domain_studying_career():
    """Test the diversity over two equals agents"""
    agents = [
        Agent("One", [
            Attribute("competences.u_active", QN, [0, 1], 0.0),
            Attribute("competences.u_read", QN, [0, 1], 0.0),
            Attribute("competences.u_essay", QN, [0, 1], 0.0),
            Attribute("competences.u_org", QN, [0, 1], 0.0),
            Attribute("competences.u_balance", QN, [0, 1], 0.0),
            Attribute("competences.u_assess", QN, [0, 1], 0.0),
            Attribute("competences.u_theory", QN, [0, 1], 0.0),
            Attribute("competences.u_pract", QN, [0, 1], 0.0)
            ]),
        Agent("Two", [
            Attribute("competences.u_active", QN, [0, 1], 0.4),
            Attribute("competences.u_read", QN, [0, 1], 0.4),
            Attribute("competences.u_essay", QN, [0, 1], 0.4),
            Attribute("competences.u_org", QN, [0, 1], 0.4),
            Attribute("competences.u_balance", QN, [0, 1], 0.4),
            Attribute("competences.u_assess", QN, [0, 1], 0.4),
            Attribute("competences.u_theory", QN, [0, 1], 0.4),
            Attribute("competences.u_pract", QN, [0, 1], 0.4)
            ])
    ]
    reqattr = [
        Attribute("competences.u_active", QN, [0, 1]),
        Attribute("competences.u_read", QN, [0, 1]),
        Attribute("competences.u_essay", QN, [0, 1]),
        Attribute("competences.u_org", QN, [0, 1]),
        Attribute("competences.u_balance", QN, [0, 1]),
        Attribute("competences.u_assess", QN, [0, 1]),
        Attribute("competences.u_theory", QN, [0, 1]),
        Attribute("competences.u_pract", QN, [0, 1])
        ]
    div = diversity(agents, reqattr)
    assert  div == 1.5999999999999999


startDevelopmentEnvironment.sh
#!/bin/bash
if [ -f /.dockerenv ]; then
   echo "You can not start the development environment inside a docker container"
else
	DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
	pushd "$DIR" >/dev/null
	VERSION=$(grep --max-count=1 'version' app/main.py | awk -F '"' '{ print $2 }' | awk -F '"' '{ print $1 }')
	DOCKER_BUILDKIT=1 docker build -f docker/Dockerfile -t internetofus/profile-diversity-manager:$VERSION .
	if [ $? -eq 0 ]; then
		docker tag internetofus/profile-diversity-manager:$VERSION internetofus/profile-diversity-manager:latest
		DOCKER_BUILDKIT=1 docker build -f docker/dev/Dockerfile -t internetofus/profile-diversity-manager:dev .
		docker run --rm --name profile_diversity_manager_dev -v "${PWD}":/app -p 9000:9000 -it internetofus/profile-diversity-manager:dev /bin/bash
	fi
	popd >/dev/null
fi

app/main.py
#
# Copyright 2019 - 2022 UDT-IA, IIIA-CSIC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from enum import Enum
import json
from math import comb
import os
import re
import time

from WNS import sim_str_str, detect_lang
from diversity import Attribute, Agent, QN, QL, diversity
from fastapi import FastAPI, Body, HTTPException
from pydantic import BaseModel, Field, HttpUrl
from typing import List, Dict


# Setup Web services
app  = FastAPI(
    title="ProfileDiversityManager",
    description="Extension of the profile manager to manage the diversity.",
    version="0.2.0"
)

# Obtain information of the API
class Info(BaseModel):
    name: str = Field(...,description="Contain the name of the API.",example="wenet/profile-diversity-manager")
    apiVersion: str = Field(...,description="Contain the name of the API.",example="1.0.0")
    softwareVersion: str = Field(...,description="Contain the implementation version number of the software.",example="1.0.0")
    vendor: str = Field(...,description="Contain information of the organization that has implemented the API.",example="UDT-IA, IIIA-CSIC")
    license: str = Field(...,description="Contain information of the license of the API.",example="Apache v2")
    
# Information of an agent to obtain the diversity
class AgentData(BaseModel):
    id: str = Field(...,description="Contain the identifier of the agent.",example="1")
    quantitativeAttributes: Dict[str,float] = Field({},ge=0.0,le=1.0,description="The quantitative attributes of the agent, where the key is the name of the attribute and the value is a number on the range [0,1] that represents the attribute value.",example="{\"introvert\":0.5,\"extrovert\":0.0,\"naturalist\":1.0}")
    qualitativeAttributes: Dict[str,str] = Field({},description="The qualitative attributes of the agent, where the key is the attribute name and the value is a string with the attribute value.",example="{\"gender\":\"F\",\"civilStatus\":\"single\"}")

# Set of agents to calculate its diversity
class AgentsData(BaseModel):
    agents:List[AgentData] = Field([],min_items=2,description="A set of agents to use.")
    qualitativeAttributes: Dict[str,List[str]] = Field({},description="The name and the possible values of the qualitative attributes, where the key is the name of the attribute and the value is a list with the possible values.",example="{\"gender\":[\"M\",\"F\",\"O\"],\"civilStatus\":[\"single\",\"married\",\"divorced\",\"widow\"]}")
    quantitativeAttributes: List[str] = Field([],description="The name of the quantitative attributes",example="[\"introvert\",\"extrovert\",\"naturalist\"]")
    
# The diversity of some users
class Diversity(BaseModel):
    value: float = Field(...,ge=0.0,le=1.0,description="The diversity of a set of users.")

# The type of statistic function
class Aggregation(str, Enum):
    max = "max"
    mean = "mean"
    q75 = "q75"
    q90 = "q90"

# The data to filter the attributes
class AttributesData(BaseModel):
    source: str = Field(...,description="Source string to obtain the similarity.",example="I hate planes")
    attributes: List[str] = Field([],description="Attributes to filter",example="[\"attribute1\",\"attribute2\"]")
    aggregation: Aggregation = Field("max",description="The aggregation value to use")

# The similarity of an attribute
class AttributeSimilarity(BaseModel):
    attribute: str = Field(...,description="The name of the attribute.",example="\"attribute1\"")
    similarity: float = Field(0.0,ge=0.0,le=1.0,description="The similarity of the attribute.")

# The diversity of some users
class Similarities(BaseModel):
    similarities: List[AttributeSimilarity] = Field([],description="The similarities of the text with the attributes.")
    
@app.get(
    "/help/info",
    description="Obtain information of the extension",
    status_code=200,
    response_model=Info
)
async def get_help_info():
    
    return {
        "name": "wenet/profile-diversity-manager",
        "apiVersion": "0.1.0",
        "softwareVersion": "0.2.0",
        "vendor": "UDT-IA, IIIA-CSIC",
        "license": "Apache v2"
    } 

@app.post(
    "/calculateDiversityOf",
    description="Obtain the diversity that is formed per  a set of users",
    status_code=200,
    response_model=Diversity
)
async def post_calculate_diversity_of(data:AgentsData):

    users_diversity = 0.0
    try:
        reqattr = []
        for attrName in data.quantitativeAttributes:
            
            reqattr.append(Attribute(attrName,QN,[0,1]))
    
        for attrName in data.qualitativeAttributes:
            
            reqattr.append(Attribute(attrName,QL,data.qualitativeAttributes[attrName]))
    
        max = len(reqattr)
        if max > 0:
            
            agents = []
            for agentData in data.agents:
                agent_attributes = []
                for attrName in agentData.quantitativeAttributes:
                
                    agent_attributes.append(Attribute(attrName,QN,[0,1],agentData.quantitativeAttributes[attrName]))
        
                for attrName in agentData.qualitativeAttributes:
                
                    agent_attributes.append(Attribute(attrName,QL,data.qualitativeAttributes[attrName],agentData.qualitativeAttributes[attrName]))
        
                agents.append(Agent(agentData.id,agent_attributes))
                
            if len(agents) > 0:
                # Calculate the diversity of the agents
                users_diversity = diversity(agents,reqattr)
                users_diversity /= max # To normalize on the range [0,1]
    except:
        #Ignore exceptions
        users_diversity = 0.0                
    return {
        "value": users_diversity
    } 

@app.post(
    "/calculateSimilarityOf",
    description="Obtain the similarity between some attributes and some text",
    status_code=200,
    response_model=Similarities
)
async def post_calculate_similarity_of(data:AttributesData):
    
    attribute_similarities = []
    try:
        source_lang = detect_lang(data.source)
        pattern_camel = re.compile(r'(?<!^)(?=[A-Z])')
        pattern_word = re.compile(r'(\W|\.|_)')
        for attribute in data.attributes:
            normalized_attribute = pattern_camel.sub(' ',attribute).lower()
            normalized_attribute = pattern_word.sub(' ',normalized_attribute)
            sim = sim_str_str(data.source,normalized_attribute,source_lang,"en",data.aggregation)
            attribute_similarities.append({"attribute":attribute,"similarity":sim})
    except:
        #Ignore exceptions
        attribute_similarities = []                
        
    return {
        "similarities": attribute_similarities
    } 

